R/GRANonGRAN.R

```diff
@@ -1,86 +1,151 @@
+  repo_results(repo)[granInd, ] = ResultsRow(name = pkgname)
+
+  ##    cran_use_ok = use_cran_granbase(repo)
+  cran_use_ok = FALSE
+
+  if (cran_use_ok) {
+    ## This should give us GRANBase, switchr, and dependencies
+    res = tryCatch(
+      install.packages(
+        "GRANBase",
+        dependencies = TRUE,
+        lib = temp_lib(repo)
+      ),
+      error = function(e)
+        e
+    )
+    if (is(res, "error"))
+      cran_use_ok = FALSE
+  }
+
+  if (!cran_use_ok) {
+    ## Force switchr and GRANBase into the manifest and make them build
+    ## Don't build them if there has been no version bump
+
+    pkgs <- c("switchr",     #[1]
+              "GRANBase")    #[2]
+    old_df <- repo_results(repo)
+    mini_df <- old_df[old_df$name %in% pkgs, ]
+    old_switchr_ver <- mini_df$lastbuiltversion[mini_df$name == pkgs[1]]
+    old_gran_ver <- mini_df$lastbuiltversion[mini_df$name == pkgs[2]]
+
+    # Construct raw Github URLs for DESCRIPTION files of GRAN and switchr
+    github_user <- "gmbecker"
+    github_base_url <- "https://github.com"
+    github_raw_base <- "https://raw.githubusercontent.com"
+    repo_names <- c("switchr", #[1]
+                    "gran")    #[2]
+    repo_urls <- paste0(github_base_url, "/", github_user, "/", repo_names)
+    raw_desc_urls <- paste0(github_raw_base, "/", github_user,
+                            "/", repo_names, "/master/DESCRIPTION")
+    switchr_conn <- url(raw_desc_urls[1])
+    gran_conn <- url(raw_desc_urls[2])
+
+    # Use read.dcf to extract the live versions of GRANBase and switchr
+    curr_switchr_ver <- read.dcf(switchr_conn, fields = "Version")
+    curr_gran_ver <- read.dcf(gran_conn, fields = "Version")
+
+    # Close the open URL connections
+    close(switchr_conn)
+    close(gran_conn)
+
+    # Replace NAs, NULLs or character(0) with 0.0-0
+    check_invalid <- function(x) {
+      if (is.na(x) || is.null(x) || identical(x, character(0))) {
+        return("0.0-0")
+      } else {
+        return(x)
+      }
     }
-
-    if(!cran_use_ok) { ## force switchr and GRANBase into the manifest and make them build
-        pkgs = c("switchr", "GRANBase")
-        repo = addPkg(repo,
-                      name = pkgs,
-                      url = c("https://github.com/gmbecker/switchr",
-                              "https://github.com/gmbecker/gRAN"),
-                      type = "git", replace=TRUE)
-
-        df = repo_results(repo)
-        df[df$name %in% pkgs, "building"] = TRUE
-        df[df$name %in% pkgs, "lastbuiltversion"] = "0.0-0"
-
-        repo_results(repo) = df
+    old_switchr_ver <- check_invalid(old_switchr_ver)
+    old_gran_ver <- check_invalid(old_gran_ver)
+
+    # if old_switchr_ver is not the same as curr_switchr_ver, or
+    # if old_gran_ver is not the same as curr_gran_ver,
+    # then force them into the repo and build them
+    # Otherwise, don't force them in
+    if ((old_switchr_ver != curr_switchr_ver) || (old_gran_ver != curr_gran_ver)) {
```

Comment:
Please change this check so that the builds happen only if the "new" switchr is *later* than the old switchr version. The reverse is unlikely but we don't want to build in that case. You can use compareVersion() for this.